File: python/unified_webhook_router/providers/twilio.py
import json
from typing import Dict, Any
from urllib.parse import parse_qs
from ..types import WebhookProvider
from ..utils import compute_hmac, timing_safe_compare
import hashlib
import base64


class TwilioProvider(WebhookProvider):
    @property
    def name(self) -> str:
        return 'twilio'
    
    def identify(self, headers: Dict[str, str], body: bytes) -> bool:
        return 'x-twilio-signature' in headers
    
    def verify(self, headers: Dict[str, str], raw_body: bytes, config: Dict[str, Any]) -> bool:
        signature = headers.get('x-twilio-signature')
        if not signature or not config.get('auth_token'):
            return False
        
        # For Twilio, we need the full URL
        # Try to reconstruct from headers or use config
        proto = headers.get('x-forwarded-proto', 'https')
        host = headers.get('host', '')
        path = headers.get('x-original-url', '')
        url = config.get('webhook_url', f"{proto}://{host}{path}")
        
        # Parse the body to get parameters
        content_type = headers.get('content-type', '')
        params = {}
        
        if 'application/x-www-form-urlencoded' in content_type:
            parsed = parse_qs(raw_body.decode('utf-8'))
            params = {k: v[0] if len(v) == 1 else v for k, v in parsed.items()}
        elif 'application/json' in content_type:
            params = json.loads(raw_body.decode('utf-8'))
        
        # Sort parameters alphabetically and concatenate
        sorted_keys = sorted(params.keys())
        data = url
        
        for key in sorted_keys:
            data += key + str(params[key])
        
        # Compute HMAC-SHA1
        auth_token = config['auth_token']
        if isinstance(auth_token, str):
            auth_token = auth_token.encode('utf-8')
        
        signature_bytes = hashlib.sha1(auth_token + data.encode('utf-8')).digest()
        expected_signature = base64.b64encode(signature_bytes).decode('utf-8')
        
        return timing_safe_compare(signature, expected_signature)
    
    def extract_event_type(self, headers: Dict[str, str], payload: Any) -> str:
        # Twilio doesn't have a specific event type field
        if payload.get('MessageStatus'):
            return 'message.status'
        elif payload.get('CallStatus'):
            return 'call.status'
        elif payload.get('Body') and payload.get('From'):
            return 'message.received'
        return 'webhook'
    
    def parse_payload(self, raw_body: bytes, headers: Dict[str, str]) -> Any:
        content_type = headers.get('content-type', '')
        
        if 'application/x-www-form-urlencoded' in content_type:
            parsed = parse_qs(raw_body.decode('utf-8'))
            payload = {k: v[0] if len(v) == 1 else v for k, v in parsed.items()}
            return payload
        else:
            return json.loads(raw_body.decode('utf-8'))

File: typescript/src/providers/twilio.ts
import { WebhookProvider } from '../types';
import { computeHMAC, timingSafeCompare } from '../utils/crypto';
import * as crypto from 'crypto';

export const TwilioProvider: WebhookProvider = {
  name: 'twilio',
  
  identify: (headers: Record<string, string>) => {
    return 'x-twilio-signature' in headers;
  },
  
  verify: (headers: Record<string, string>, rawBody: Buffer, config: any) => {
    const signature = headers['x-twilio-signature'];
    if (!signature || !config.authToken) {
      return false;
    }
    
    // For Twilio, we need the full URL which should be passed in config
    const url = headers['x-forwarded-proto'] + '://' + headers['host'] + headers['x-original-url'] || config.webhookUrl || '';
    
    // Parse the body to get parameters
    const contentType = headers['content-type'] || '';
    let params: Record<string, string> = {};
    
    if (contentType.includes('application/x-www-form-urlencoded')) {
      const searchParams = new URLSearchParams(rawBody.toString('utf8'));
      searchParams.forEach((value, key) => {
        params[key] = value;
      });
    } else if (contentType.includes('application/json')) {
      params = JSON.parse(rawBody.toString('utf8'));
    }
    
    // Sort parameters alphabetically and concatenate
    const sortedKeys = Object.keys(params).sort();
    let data = url;
    
    for (const key of sortedKeys) {
      data += key + params[key];
    }
    
    // Compute HMAC-SHA1
    const expectedSignature = crypto
      .createHmac('sha1', config.authToken)
      .update(data)
      .digest('base64');
    
    return timingSafeCompare(signature, expectedSignature);
  },
  
  extractEventType: (headers: Record<string, string>, payload: any) => {
    // Twilio doesn't have a specific event type field
    // We can use the webhook type based on the payload
    if (payload.MessageStatus) {
      return 'message.status';
    } else if (payload.CallStatus) {
      return 'call.status';
    } else if (payload.Body && payload.From) {
      return 'message.received';
    }
    return 'webhook';
  },
  
  parsePayload: (rawBody: Buffer, headers: Record<string, string>) => {
    const contentType = headers['content-type'] || '';
    
    if (contentType.includes('application/x-www-form-urlencoded')) {
      const params = new URLSearchParams(rawBody.toString('utf8'));
      const payload: any = {};
      for (const [key, value] of params) {
        payload[key] = value;
      }
      return payload;
    } else {
      return JSON.parse(rawBody.toString('utf8'));
    }
  }
};